### ISSUE TITLE
Review HPWH object structure - HVACDiagram fails with heat pump water heater example

### ISSUE THREAD

**Issue by mjwitte** [NREL/EnergyPlus#5055]
Helpdesk ticket 10646
Use example file HeatPumpWaterHeater.idf.  HVACDiagram reports:
Error: No node connection found after the following object: ZONE4BHPWHFAN
Likely cause is the way nodes get implied inside the HPWH connections.  There may be a solution within the getonlysinglenode calls to generate a working bnd output for HVACDiagram to draw.
This example file and user's file added as defect files.

**Comment by kbenne** [comment]
> Likely cause is the way nodes get implied inside the HPWH connections

Yes this causes hoops for OpenStudio as well. In OpenStudio base class nomenclature we have HeatPumpWaterHeater as a subclass of ZoneHVACComponent masquerading about with a dual personality of HVACComponent (something that goes on a loop). We avoid multiple inheritance by factoring out the tank itself as the actual plant component and then reassembling all of the node connections in the ism->idf translator, but it is not ideal.

Similar to https://github.com/NREL/EnergyPlusDevSupport/commit/8bc9900ac78ef5c462f1d8a7fc7e7e3a7109c8ca#commitcomment-12254432 ?

**Comment by mjwitte** [comment]
@kbenne I think this is different.  On the air side, there is an implied mixer (if I remember correctly) that is not explicitly called out as a component in the HPWH input.

**Comment by kbenne** [comment]
Fair enough. I think you are right about the implied mixer on HPWH. Even though I just worked through this in OS the details leave my brain quickly. In any case I agree it is different, but the hoops are still similar enough from an OS point of view. Pardon my interruption while I grumble a little...

**Comment by mjwitte** [comment]
@kbenne This is a valid concern that we should think about changing.  The timing is a bit unfortunate as the recent residential HPWH upgrades imposed a significant transition rule, and this might incur yet another.  But if it's causing fits for interfaces, we should figure out a better approach.j

**Comment by kbenne** [comment]
at this point changing it would cause more fits for at least one interface :) and I don't exactly have a better alternative formulated yet. I'd have to review my pain points if you really want to get into it more.

**Comment by kbenne** [comment]
Taking a quick look to remind myself here is what comes to mind.

The water side node names are duplicated on the HeatPumpWaterHeater object and the tank itself.

https://github.com/NREL/EnergyPlus/blob/develop/testfiles/HeatPumpWaterHeater.idf#L2147
https://github.com/NREL/EnergyPlus/blob/develop/testfiles/HeatPumpWaterHeater.idf#L2213

While the heat pump also references the air side zone nodes 
https://github.com/NREL/EnergyPlus/blob/develop/testfiles/HeatPumpWaterHeater.idf#L2152

thus making the component fit the bill for both ZoneHVACComponent and HVACComponent. We have a derived type from HVACComponent called WaterToAirComponent but that is for coils, not zone equipment so I don't think that is really appropriate.

In any case in OpenStudio we consider the tank the actual plant supply component (with node connections) and leave the HeatPumpWaterHeater as pure zone HVAC, then match up the nodes names in translation. This is probably more similar to how things like fan coils are done in OS AND E+, the difference being that the "inner" thing is a plant supply component in this case as opposed to a demand like a coil.

Based on the OS approach, in addition to needing to patch up the water side node names on the HeatPumpWaterHeater object, we also have to do some shenanigans to make sure it is the heat pump that ends up on the plant operation schemes as opposed to the tank like OpenStudio wants to do when it encounters a plant component.

These might be issues specific to how OS has chosen to select base classes I don't know.  But our bases are pretty rational I think.  And one thing about OO, it forces you to be pretty consistent unless you want to experience pain...

**Comment by kbenne** [comment]
I almost hate to point out this abomination, but here is some ugly code to begin with in OpenStudio, regardless of the issues of the heat pump.  And [here](https://github.com/NREL/OpenStudio/blob/8a380637320bbebadaa063d05dbeb0091fec443a/openstudiocore/src/energyplus/ForwardTranslator/ForwardTranslatePlantLoop.cpp#L293) you can see the shenanigans. 

And more [shenanigans](https://github.com/NREL/OpenStudio/blob/8a380637320bbebadaa063d05dbeb0091fec443a/openstudiocore/src/energyplus/ForwardTranslator/ForwardTranslatePlantLoop.cpp#L488).

The later is related to code that tries to build up default operation schemes. Some of this I am cleaning up now because we are giving users explicit control over operation schemes.  But the default logic will remain in some form for users that don't want to think about it.

The first link is in the context of code that builds up the branches, since OS does not have the notion of branches in favor of a generalized directed graph.

**Comment by kbenne** [comment]
@mjwitte gee you crack the door open and see what I do. :)

**Comment by mjwitte** [comment]
@kbenne Always good to let vent the frustrations . . .
